File: src/main/java/editor/gui/editor/IncludeExcludePanel.java
package editor.gui.editor;

import java.awt.Color;
import java.awt.Dimension;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import javax.swing.BoxLayout;

import editor.collection.deck.CategorySpec;
import editor.database.card.Card;
import editor.gui.generic.ScrollablePanel;
import editor.gui.generic.TristateCheckBox;
import editor.gui.generic.TristateCheckBox.State;

/**
 * This class represents a panel that displays a list of categories with boxes next to them indicating
 * card inclusion in each of them.  A check mark means all cards it was given are included, a "mixed" icon
 * indicates some of the cards are included, and an empty box indicates none are included.  Boxes can be
 * selected to toggle inclusion/exclusion of cards (mixed cannot be selected manually).  Card inclusion is
 * not changed by the panel, but it provides information that allows it to be externally.
 *
 * @author Alec Roelke
 */
public class IncludeExcludePanel extends ScrollablePanel
{
    /**
     * Maximum amount of rows to display in a scroll pane.
     */
    private static final int MAX_PREFERRED_ROWS = 10;

    /**
     * Categories and their corresponding check boxes.
     */
    private Map<CategorySpec, TristateCheckBox> categoryBoxes;
    /**
     * List of cards to display inclusion for.
     */
    private Collection<Card> cards;
    /**
     * Preferred viewport height of this panel.
     */
    private int preferredViewportHeight;

    /**
     * Create a new IncludeExcludePanel showing inclusion of the given cards in the given
     * categories.
     *
     * @param categories categories to display
     * @param c          cards to show inclusion for
     */
    public IncludeExcludePanel(List<CategorySpec> categories, Collection<Card> c)
    {
        super(TRACK_WIDTH);
        setLayout(new BoxLayout(this, BoxLayout.Y_AXIS));
        setBackground(Color.WHITE);
        categoryBoxes = new HashMap<>();
        cards = c;
        preferredViewportHeight = 0;

        for (CategorySpec category : categories)
        {
            TristateCheckBox categoryBox = new TristateCheckBox(category.getName());
            long matches = cards.stream().filter(category::includes).count();
            if (matches == 0)
                categoryBox.setState(State.UNSELECTED);
            else if (matches < cards.size())
                categoryBox.setState(State.INDETERMINATE);
            else
                categoryBox.setState(State.SELECTED);
            categoryBox.setBackground(Color.WHITE);
            add(categoryBox);
            categoryBoxes.put(category, categoryBox);
            preferredViewportHeight = Math.min(preferredViewportHeight + categoryBox.getPreferredSize().height, categoryBox.getPreferredSize().height * MAX_PREFERRED_ROWS);
        }
    }

    /**
     * Create a new IncludeExcludePanel for a single card.
     *
     * @param categories categories to show inclusion for
     * @param card       card to show inclusion for
     */
    public IncludeExcludePanel(List<CategorySpec> categories, Card card)
    {
        this(categories, Collections.singletonList(card));
    }

    /**
     * Get the cards that were selected for inclusion.  Ignore cards that were already included
     * from the start.
     *
     * @return a Map containing the cards that were selected for inclusion and the new categories
     * they were included in.
     */
    public Map<Card, Set<CategorySpec>> getIncluded()
    {
        var included = new HashMap<Card, Set<CategorySpec>>();
        for (Card card : cards)
            for (CategorySpec category : categoryBoxes.keySet())
                if (categoryBoxes.get(category).getState() == State.SELECTED && !category.includes(card))
                    included.compute(card, (k, v) -> {
                        if (v == null)
                            v = new HashSet<>();
                        v.add(category);
                        return v;
                    });
        return included;
    }

    /**
     * Get the cards that were deselected for exclusion.  Ignore cards that were already excluded
     * from the start.
     *
     * @return a Map containing the cards that were deselected for exclusion and the categories
     * they should be excluded from.
     */
    public Map<Card, Set<CategorySpec>> getExcluded()
    {
        var excluded = new HashMap<Card, Set<CategorySpec>>();
        for (Card card : cards)
            for (CategorySpec category : categoryBoxes.keySet())
                if (categoryBoxes.get(category).getState() == State.UNSELECTED && category.includes(card))
                    excluded.compute(card, (k, v) -> {
                        if (v == null)
                            v = new HashSet<>();
                        v.add(category);
                        return v;
                    });
        return excluded;
    }

    /**
     * {@inheritDoc}
     * The preferred viewport size of this IncludeExcludePanel is the size of its contents
     * up to {@link IncludeExcludePanel#MAX_PREFERRED_ROWS}.
     */
    @Override
    public Dimension getPreferredScrollableViewportSize()
    {
        if (categoryBoxes.isEmpty())
            return getPreferredSize();
        else
        {
            Dimension size = getPreferredSize();
            size.height = preferredViewportHeight;
            return size;
        }
    }
}

File: src/main/scala/editor/gui/editor/IncludeExcludePanel.scala
package editor.gui.editor

import editor.collection.Categorization
import editor.database.card.Card
import editor.gui.generic.ScrollablePanel
import editor.gui.generic.TristateCheckBox
import editor.gui.generic.TristateCheckBoxState

import java.awt.Color
import javax.swing.BoxLayout

/**
 * Panel allowing the user to edit the category inclusion of cards.  It contains a list of check boxes whose labels
 * are the names of categories. A checked box indicates that all cards in the list are included in the category, an
 * unchecked box indicates none of them are, and a partially-checked (square icon) box indicates that some are. The
 * user can only fully check or uncheck boxes.
 * 
 * @constructor create a new panel for editing the inclusion of a list of cards for a list of categories
 * @param categories categories to edit
 * @param cards cards to include or exclude
 * 
 * @author Alec Roelke
 */
class IncludeExcludePanel(categories: Seq[Categorization], cards: Seq[Card]) extends ScrollablePanel(ScrollablePanel.TrackWidth) {
  private val MaxPreferredRows = 10

  setLayout(BoxLayout(this, BoxLayout.Y_AXIS))
  setBackground(Color.WHITE)

  private val categoryBoxes = categories.map((category) => {
    val matches = cards.count(category)
    val categoryBox = TristateCheckBox(category.name, 
      if (matches == 0)
        TristateCheckBoxState.Unselected
      else if (matches < cards.size)
        TristateCheckBoxState.Indeterminate
      else
        TristateCheckBoxState.Selected
    )
    categoryBox.setBackground(Color.WHITE)
    category -> categoryBox
  }).toMap
  categories.foreach((c) => add(categoryBoxes(c)))

  /**
   * Get the categories that should have cards included in them. Categories are not actually updated.
   * @return a mapping of cards onto the categories they should be included in
   */
  def toInclude = cards.map((c) => c -> categoryBoxes.collect{ case (category, box) if (box.state == TristateCheckBoxState.Selected && !category(c)) => category }.toSet).toMap

  /**
   * Get the categories that should have cards excluded from them. Categories are not actually updated.
   * @return a mapping of cards onto the categories they should be excluded from
   */
  def toExclude = cards.map((c) => c -> categoryBoxes.collect{ case (category, box) if (box.state == TristateCheckBoxState.Unselected && category(c)) => category }.toSet).toMap

  /** @return copies of the provided categories, updated to include or exclude cards as selected by the user. */
  def updates = {
    val inc = toInclude
    val exc = toExclude
    categories.collect{ case category if inc.values.exists(_.contains(category)) || exc.values.exists(_.contains(category)) =>
      category ++ inc.collect{ case (card, in) if in.contains(category) => card } -- exc.collect{ case (card, ex) if ex.contains(category) => card }
    }
  }

  override def getPreferredScrollableViewportSize = {
    val size = getPreferredSize
    if (!categoryBoxes.isEmpty)
      size.height = Math.min(
        categoryBoxes.map{ case (_, b) => b.getPreferredSize.height }.fold(0)(_ + _),
        categoryBoxes.headOption.map{ case (_, b) => b.getPreferredSize.height*MaxPreferredRows }.getOrElse(0)
      )
    size
  }
}
